In [1]:
#importing necessary libraries
import numpy as np
import pandas as pd
import keras

Using TensorFlow backend.


In [2]:
#fetching the data into a dataframe
data = pd.read_csv("concrete_data.csv")
data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
#checking whether the data is clean or not
data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [4]:
#since the data is clean their is no further requirement of cleaning the data
#let us split the data in predictors and the target value
predictors = data.iloc[:,:8]
target = data['Strength']

In [5]:
#Normalizing the data
predictors = (predictors - predictors.mean()) / predictors.std()
predictors.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [6]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [7]:
#importing necessary packages for building the model
from keras.models import Sequential
from keras.layers import Dense

In [9]:
#Defing the regression_model
number_of_inputs = predictors.shape[1]
def regression_model():
    
    model = Sequential()
    
    #Adding the layes pof the model
    model.add(Dense(10, activation = 'relu', input_shape = (number_of_inputs,)))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(1))
    
    #compiling the model with adam optimizer
    model.compile(optimizer = 'adam', loss = 'mean_squared_error') 
    return model

In [10]:
model = regression_model()

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [12]:
MSE=[]
for i in range(50):
    X_train,X_test,y_train,y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, validation_data = (X_test, y_test), epochs = 100, verbose = 0)        
    mse = model.evaluate(X_test, y_test, verbose=0)
    print("Iteration:",i+1,"Mean Squared Error:",mse)
    MSE.append(mse)

Iteration: 1 Mean Squared Error: 102.78065253692924
Iteration: 2 Mean Squared Error: 81.06610132112472
Iteration: 3 Mean Squared Error: 41.883077584424065
Iteration: 4 Mean Squared Error: 36.200669421347214
Iteration: 5 Mean Squared Error: 35.12225026991761
Iteration: 6 Mean Squared Error: 35.194297062154725
Iteration: 7 Mean Squared Error: 36.18844394622111
Iteration: 8 Mean Squared Error: 25.893790461098877
Iteration: 9 Mean Squared Error: 29.71319386256937
Iteration: 10 Mean Squared Error: 28.198383158464647
Iteration: 11 Mean Squared Error: 24.84241116471275
Iteration: 12 Mean Squared Error: 22.07739599160006
Iteration: 13 Mean Squared Error: 30.895825818132813
Iteration: 14 Mean Squared Error: 30.89300159343238
Iteration: 15 Mean Squared Error: 23.566079679816287
Iteration: 16 Mean Squared Error: 23.959507695293734
Iteration: 17 Mean Squared Error: 25.83068013114065
Iteration: 18 Mean Squared Error: 25.990299345220176
Iteration: 19 Mean Squared Error: 26.50855546476
Iteration: 20 

In [13]:
mse = np.asarray(MSE)
print("Mean of Mean Squared Errors : ",np.mean(mse))
print("Standard Deviation of Mean Squared Errors : ",np.std(mse))

Mean of Mean Squared Errors :  26.782196629163124
Standard Deviation of Mean Squared Errors :  14.632180209774992


In STEP-C the mean of the Mean Squared Error was 38.04857305372417, where as in STEP-D the mean of the Mean Squared Error is 26.782196629163124. 

We can see an DECREASE in the mean.